### PR TITLE
blog: Add parsing .DS_Store files post

### PR DESCRIPTION
**Pull request by SkypLabs**
Hi,

Following https://github.com/zaproxy/zaproxy/issues/30#issuecomment-1413438603 and a discussion I had with @psiinon and @kingthorin on Mastodon, here is the blog post regarding parsing `.DS_Store` files with ZAP.

The post is not finished yet but I'm creating this draft PR to let you share your feedback along the way.

Thanks!

### CODE REVIEW

**Comment by SkypLabs** [comment]
I finally managed to find the time to finish this blog post!

I look forward to reading your feedback.

**Comment by kingthorin** [comment]
Thanks! I’ll try to review in the next few days.

File: site/content/blog/2023-09-04-ds-store-parsing/index.md [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+type: post
+tags:
+- blog
+- recon
+- spider
+date: "2023-09-04"
+authors:
+    - skyper
+---
+macOS has been using [`.DS_Store`][wikipedia-ds-store] hidden files for many
```
Comment (kingthorin) [inline site/content/blog/2023-09-04-ds-store-parsing/index.md:14]:
```suggestion
---

macOS has been using [`.DS_Store`][wikipedia-ds-store] hidden files for many
```

File: site/content/blog/2023-09-04-ds-store-parsing/index.md [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+stored in the Git repository of a website.
+
+Discovering and reading the contents of `.DS_Store` files present on a website
+can prove to be a valuable asset when conducting web security audits and
+penetration tests. This has led me to the idea of integrating this
+reconnaissance task as part of the ZAP Spider tool, enabling the process to be
+fully automated and saving time during security engagements. But first things
+first, one needs to parse the internal structure of such files.
+
+## Parsing .DS_Store files
```
Comment (kingthorin) [inline site/content/blog/2023-09-04-ds-store-parsing/index.md:32]:
```suggestion
## Parsing .DS_Store Files
```

File: site/content/blog/2023-09-04-ds-store-parsing/index.md [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+The specifications detailing the internal structure of `.DS_Store` files have
+never been made public, but [thanks to community-driven
+efforts][0day-parsing-ds-store], most parts have been reverse engineered,
+enabling people to create custom parsers. Nevertheless, writing a parser for
+`.DS_Store` files is not a trivial task given the complexity of the format. This
+is why not much implementations can be found online, and most of them are in
+Python, Go and Perl only.
```
Comment (kingthorin) [inline site/content/blog/2023-09-04-ds-store-parsing/index.md:40]:
```suggestion
is why few implementations can be found online, and most of them are in
Python, Go, or Perl only.
```

File: site/content/blog/2023-09-04-ds-store-parsing/index.md [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+            self.image_height = self._io.read_u2le()
+            self.flags = self._io.read_u1()
+            self.bg_color_index = self._io.read_u1()
+            self.pixel_aspect_ratio = self._io.read_u1()
+```
+
+Note that the code produced relies on an external dependency from the Kaitai
+Struct project, so the result is not standalone.
+
+In comparison, the internal structure of a `.DS_Store` file is immensity more
```
Comment (kingthorin) [inline site/content/blog/2023-09-04-ds-store-parsing/index.md:150]:
```suggestion
In comparison, the internal structure of a `.DS_Store` file is immensely more
```

File: site/content/blog/2023-09-04-ds-store-parsing/index.md [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+[B-trees][wikipedia-b-tree], with each B-tree referred to as a "directory" in
+this file format's terminology. Each directory consists of blocks (where a block
+represents a node in a B-tree), and each block consists of records. Each record
+contains a filename and a type that defines the purpose of its value.
+
+It took me some time to write a functional Kaitai Struct format specification
+for the `.DS_Store` files, but the result is now publicly accessible in the
+[Kaitai Struct's Format Gallery][kaitai-struct-format-gallery].
+
+## Probing .DS_Store files with ZAP
```
Comment (kingthorin) [inline site/content/blog/2023-09-04-ds-store-parsing/index.md:166]:
```suggestion
## Probing .DS_Store Files with ZAP
```

File: site/content/blog/2023-09-04-ds-store-parsing/index.md (hunk 4 above)
Comment (SkypLabs) [reply site/content/blog/2023-09-04-ds-store-parsing/index.md]: Good catch! Thanks.

File: site/content/blog/2023-09-04-ds-store-parsing/index.md (hunk 3 above)
Comment (SkypLabs) [reply site/content/blog/2023-09-04-ds-store-parsing/index.md]: Agreed. It reads better this way. Thanks.

File: site/data/authors.yml [hunk 6]
```diff
@@ -36,32 +36,32 @@ david:
 jordan:
   name: Jordan GS
   picture: /img/authors/zap_400x400.png
-  twitter: 
+  twitter:
```
Comment (thc202) [inline site/data/authors.yml:39]: Would be good to revert these changes.

**Comment by thc202** [comment]
The date of blog will have to be updated.

File: site/data/authors.yml (hunk 6 above)
Comment (SkypLabs) [reply site/data/authors.yml:39]: My IDE (Vim) is configured to remove all trailing space characters upon saving. I thought it wouldn't be a problem to have them removed but I can revert it if you prefer.

File: site/content/blog/2023-09-04-ds-store-parsing/index.md [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+For each directory (B-tree) in the `.DS_Store` file, the master block is read to
+obtain the location of the root block. The B-tree is then traversed recursively
+from its root block ([see on
+GitHub](https://github.com/zaproxy/zap-extensions/blob/5e078b1770c888a1bd6071e2a314d86c3128ddb0/addOns/spider/src/main/java/org/zaproxy/addon/spider/parser/DsStoreParser.java#L64-L78)):
+
+```java
+for (MasterBlockRef masterBlockRef : dsStore.buddyAllocatorBody().directories()) {
+    // Each B-tree directory has one master block comprising metadata.
+    MasterBlock masterBlock = masterBlockRef.masterBlock();
+    getLogger().debug("Rcords: {}", masterBlock.numRecords());
```
[the remark this answers is not in the document]
Comment (SkypLabs) [reply site/content/blog/2023-09-04-ds-store-parsing/index.md:215]: The typo is present in the original source code, so I didn't fix it in my code snippet to keep it accurate.

File: site/content/blog/2023-09-04-ds-store-parsing/index.md [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+represents a node in a B-tree), and each block consists of records. Each record
+contains a filename and a type that defines the purpose of its value.
+
+It took me some time to write a functional Kaitai Struct format specification
+for the `.DS_Store` files, but the result is now publicly accessible in the
+[Kaitai Struct's Format Gallery][kaitai-struct-format-gallery].
+
+## Probing .DS_Store Files with ZAP
+
+ZAP has a tool called the Spider which enables to discover new resources on a
```
[the remark this answers is not in the document]
Comment (SkypLabs) [reply site/content/blog/2023-09-04-ds-store-parsing/index.md:169]: Both are synonyms here but I don't mind changing it.

File: site/content/blog/2023-09-04-ds-store-parsing/index.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (thc202) [reply site/content/blog/2023-09-04-ds-store-parsing/index.md:215]: Typo is being fixed.

File: site/data/authors.yml (hunk 6 above)
Comment (thc202) [reply site/data/authors.yml:39]: The problem is not removing trailing spaces, the problem is doing that in the same commit/PR as other changes.

File: site/content/blog/2023-09-04-ds-store-parsing/index.md [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+website. Basically, it works by scanning a first web page for references to
```
[the remark this answers is not in the document]
Comment (SkypLabs) [reply site/content/blog/2023-09-04-ds-store-parsing/index.md:170]: I don't mind changing it.

@kingthorin What do you think?

[the remark this answers is not in the document]
Comment (SkypLabs) [reply site/content/blog/2023-09-04-ds-store-parsing/index.md]: Done in d5d3c96.

File: site/content/blog/2023-09-04-ds-store-parsing/index.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (SkypLabs) [reply site/content/blog/2023-09-04-ds-store-parsing/index.md:215]: OK thanks. I will wait until https://github.com/zaproxy/zap-extensions/pull/4874 gets merged to update the permalink to the source code accordingly.

File: site/data/authors.yml (hunk 6 above)
Comment (SkypLabs) [reply site/data/authors.yml]: Makes sense, I understand. I like to have one commit per change too. I just thought it didn't matter for trailing spaces.

Done in 3a8e842.

**Review by kingthorin — APPROVED** [review]

File: site/data/authors.yml (hunk 6 above)
Comment (thc202) [reply site/data/authors.yml]: Could you create a PR?

Comment (SkypLabs) [reply site/data/authors.yml]: You mean for removing the trailing spaces? Sure, of course.

Sorry I misunderstood that you were interested in doing so but in another PR. I thought you just wanted to let it as it is for now.

Comment (thc202) [reply site/data/authors.yml]: Yes, that would be great!

File: site/content/blog/2023-09-04-ds-store-parsing/index.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (SkypLabs) [reply site/content/blog/2023-09-04-ds-store-parsing/index.md]: https://github.com/zaproxy/zap-extensions/pull/4874 got merged so I fixed the typo in the code and updated the related permalink in 5d37cbb.

**Comment by SkypLabs** [comment]
> The date of blog post will have to be updated.

As all change requests have been addressed so far, I updated the date of the blog post in [df92c03](https://github.com/zaproxy/zaproxy-website/pull/2022/commits/df92c037658a4d5cd5b31f4e729bce798befd34a).

File: site/data/authors.yml (hunk 6 above)
Comment (SkypLabs) [reply site/data/authors.yml]: PR created: https://github.com/zaproxy/zaproxy-website/pull/2100.

**Comment by psiinon** [comment]
@SkypLabs if you can add this to your PR (and link to it in the blog post) then I think we could be ready to go?

**Comment by SkypLabs** [comment]
> @SkypLabs if you can add this to your PR (and link to it in the blog post) then I think we could be ready to go?

Done in [ab8becb](https://github.com/zaproxy/zaproxy-website/pull/2022/commits/ab8becb3f03bb0cbab6a7e261a78b682a86593d0).

**Review by kingthorin — APPROVED** [review]

**Review by psiinon — APPROVED** [review]

**Comment by SkypLabs** [comment]
Great, thanks!

**Comment by psiinon** [comment]
@SkypLabs many thanks!
btw https://0day.work/parsing-the-ds_store-file-format/ seems to be down :/ Dont suppose you have an alternative link?

**Comment by SkypLabs** [comment]
>  btw https://0day.work/parsing-the-ds_store-file-format/ seems to be down :/ Dont suppose you have an alternative link?

Too bad, it was a great article :disappointed: I don't have an alternative link, but we can still read it on Web Archive: https://web.archive.org/web/20230607153117/https://0day.work/parsing-the-ds_store-file-format/.
